toolchain/check/import_ref.cpp

```diff
@@ -676,6 +676,74 @@ static auto AddImportIRInst(ImportContext& context, SemIR::InstId inst_id)
       {.ir_id = context.import_ir_id(), .inst_id = inst_id});
 }
 
+// Sets the constant value for an instruction and returns it. The constant value
+// is computed if not explicitly specified.
+static auto ComputeOrSetConstantValue(
+    Context& context, SemIR::InstId inst_id, SemIR::Inst inst,
+    std::optional<SemIR::ConstantId> const_id = std::nullopt)
+    -> SemIR::ConstantId {
+  if (!const_id) {
+    const_id = TryEvalInstUnsafe(context, inst_id, inst);
+  }
+  if (const_id->has_value()) {
+    CARBON_VLOG_TO(context.vlog_stream(), "Constant: {0} -> {1}\n", inst,
+                   context.constant_values().GetInstId(*const_id));
+  }
+  context.constant_values().Set(inst_id, *const_id);
+  return *const_id;
+}
+
+// Adds an imported instruction. The constant value of the instruction can be
+// specified if it is known, otherwise it will be computed.
+//
+// Normally `AddImportedConstant` should be used to create the result of
+// resolving a constant, for example by calling `ResolveAs*`. However, we
+// sometimes need to create a new instruction, not just obtain a constant value,
+// such as when importing a declaration or a pattern. In that case, this
+// function should be used to create it.
+//
+// Do not use `AddInst*`, as it will add symbolic constants to the eval block of
+// whatever generic the import happens within.
+template <typename InstT>
+static auto AddImportedInst(
+    ImportContext& context, SemIR::InstId import_inst_id, InstT inst,
+    std::optional<SemIR::ConstantId> const_id = std::nullopt) -> SemIR::InstId {
```

Comment:
With the way this is used, could this be using `SemIR::ConstantId::None` in place of `std::optional` + `std::nullopt`? It looks like `AddPlaceholderImportedInst` might be the only spot that actually passes `None` right now: what if that didn't defer to `AddImportedInst`, meaning `ComputeOrSetConstantValue` could also always eval?

Reply:
Rearranged somewhat as suggested: `AddImportedInst` now calls `AddPlaceholderImportedInst`, and the latter doesn't set a constant value at all.